### PR TITLE
Add distributed training support

### PR DESCRIPTION
**Pull request by sfvaroglu**
## Description ##
Add distributed training support 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: python/mxnet/module/module.py [hunk 1]
```diff
@@ -502,6 +502,9 @@ def init_optimizer(self, kvstore='local', optimizer='sgd',
         batch_size = self._exec_group.batch_size
         if kvstore and 'dist' in kvstore.type and '_sync' in kvstore.type:
             batch_size *= kvstore.num_workers
+        if kvstore and 'ngraph' in kvstore.type:
+            batch_size *= kvstore.num_workers
+            optimizer_params['learning_rate'] *= kvstore.num_workers
```
Comment (mbrookhart) [inline python/mxnet/module/module.py]: Can we leave this as a user-controlled change to learning rate instead of automating it?

File: src/ngraph/ngraph_emitter.cc [hunk 2]
```diff
@@ -1038,11 +1038,22 @@ void Emitter::CreateLayerOps() {
     const NgraphNodePtr one = makeConstant(dtype, shape, 1.0f);
 
     NgraphNodePtr scale_grad;
+
+    #if MXNET_USE_NGRAPH_DISTRIBUTED
+    NgraphNodePtr ng_grad = std::make_shared<ngraph::op::AllReduce>(grad);
```
Comment (mbrookhart) [inline src/ngraph/ngraph_emitter.cc]: Let's de-constify grad so you can simply overwrite it here and avoid the code copy.

File: src/ngraph/ngraph_emitter.cc [hunk 3]
```diff
@@ -1066,11 +1077,22 @@ void Emitter::CreateLayerOps() {
...
```
Comment (mbrookhart) [inline src/ngraph/ngraph_emitter.cc]: same here, de-constify grad to avoid the code copy

File: src/kvstore/kvstore_ngraph.h [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+#include <mxnet/kvstore.h>
+#include <mpi.h>
+
+namespace mxnet {
+namespace kvstore {
+
+/**
+ * \brief store data in local machine using nGraph
+ */
+class KVStoreNGRAPH : public KVStoreLocal {
```
Comment (wenzhe-nrv) [inline src/kvstore/kvstore_ngraph.h]: I wonder if it should be base on `public KVStore`. does Local kvstore only support single machine, which is what nccl kvstore based on? Dist sync allreduce uses `KVStore` (MX DO PR).

File: python/mxnet/module/module.py (hunk 1 above)
Comment (sfvaroglu) [reply python/mxnet/module/module.py]: Sure.

File: src/ngraph/ngraph_emitter.cc (hunk 2 above)
Comment (wenzhe-nrv) [inline src/ngraph/ngraph_emitter.cc]: the name `ng_grad` confused me a little bit. Does the `ng` stands for `ngraph`? 
I wonder if we could do reassign like this:
`grad = std::make_shared<ngraph::op::AllReduce>(grad);`? Then the snippet below could be reused too.

File: src/kvstore/kvstore_ngraph.h (hunk 4 above)
Comment (wenzhe-nrv) [reply src/kvstore/kvstore_ngraph.h]: So far as I understand, I think it doesn't matter since the real `push`/allreduce happens in the emitter. So I would change this to `public KVStore` for clarity.

Comment (sfvaroglu) [reply src/kvstore/kvstore_ngraph.h]: Our implementation is based on local kvstore. When a model is run with `mpirun`, each process gets its own kvstore. The communication takes place at the nGraph level. The other kvstore approaches are integrated with communication libraries directly.

File: NGRAPH_README.md [hunk 5]
```diff
@@ -154,6 +154,18 @@
...
+## Distributed training
+MPI is required for multi-CPU support. Download Open MPI from [here](https://www.open-mpi.org/).
+
+`USE_NGRAPH_DISTRIBUTED` must be set to (exactly) `1` in order for MXNet to build with nGraph distributed support.
+
+Here's an example to run ResNet-50 on two nodes:
+
+``` sh
+export MXNET_NGRAPH_GLUON=1
+mpirun -map-by node -x MXNET_NGRAPH_GLUON -x LD_LIBRARY_PATH -hostfile hosts -np 2 python MXNET_ROOT/example/image-classification/train_cifar10.py --network resnet --num-layers 50 --kv-store ngraph
```
Comment (ashokei) [inline NGRAPH_README.md]: does users need to do any preparation like copying code/data to all nodes , into same dir structure.

Comment (sfvaroglu) [reply NGRAPH_README.md]: Either copying the data or using NFS would work. We populate the number of workers in the run and then rely on the data iterator to split the data across nodes.

**Comment by sfvaroglu** [comment]
@ashokei Yes, each node does Allreduce. It's SPMD model.
@wenzhe-nrv I think your requests should be covered. Please let me know if you have any further requests.

**Comment by wenzhe-nrv** [comment]
Approved. 
One follow up: since `push`, `pull`, etc is not implemented in ngraph kvstore but inherent from kvstore local, should we define those methods that throw some warning? We could discuss later.

**Comment by sfvaroglu** [comment]
TESTNOW
